### PR TITLE
Full Backend Integration - Table Generator (4/24)

### PR DESCRIPTION
**Pull request by aidenhuynh**
## Backend Integration

* When adding a new class to localStorage, the user is prompted to import an existing class tied to their account on the backend, creating a copy of it locally

Here is the main function for connecting with the backend. It is async so that when it is called using `await` we can return the user's classes if there are any. If the user isn't signed in, it will return an empty array that is checked for in other parts of the code.

```js
async function fetchClassList() {
    // Create array of classes
    var classes = []

    // Fetch backend
    try {
        const response = await fetch(url + '/api/class_period/dashboard', {
            method: 'GET',
            mode: 'cors',
            cache: 'no-cache',
            credentials: 'include',
            headers: {
                "content-type": "application/json",
            },
        })

        // Error check
        if (!response.ok) {
            throw new Error('Network response was not ok')
        }

        // Convert response to JSON, which contains classes that user is the leader of
        const data = await response.json()
        console.log(JSON.stringify(data))
        var classList = data.leader

        // For each class of the user
        for (let classData of classList) {
            // Define list of students in each class
            var studentList = []

            // Push students from response into class array
            for (let student of classData.students) {
                studentList.push(student.name)
            }

            // Add each class from response into frontend class array
            classes.push({id: `class-${classData.id}`, class: studentList, name: classData.name})
        }
    } 
    // If there is an error, return an empty array
    catch (error) {
        console.error('There was a problem with the fetch operation:', error)

        // This only occurs when not signed in
        return []
    }

    return classes
}
```

On the backend, the code looks like this, returning the data tied to the user, including their classes.

```java
@GetMapping("/dashboard")
    public ResponseEntity<?> fetchBothClassData(@CookieValue("jwt") String jwtToken) {
        // checking if JWT token is missing
        if (jwtToken.isEmpty()) {
            return new ResponseEntity<>(HttpStatus.BAD_REQUEST);
        }

        // getting user data
        String userEmail = tokenUtil.getUsernameFromToken(jwtToken);
        Person existingPerson = personRepository.findByEmail(userEmail);
        if (existingPerson == null) {
            return new ResponseEntity<>(HttpStatus.BAD_REQUEST);
        }
        List<ClassPeriod> student = classPeriodDetailsService.getClassPeriodsByStudent(existingPerson);
        List<ClassPeriod> leader = classPeriodDetailsService.getClassPeriodsByLeader(existingPerson);

        // initializing storage device vrrrmmmm ERRT ERRT ERRT beeeeeep
        HashMap<String, List<ClassPeriod>> classData = new HashMap<>();

        // adding class periods to storage device brrp brrp bleeeeeeebpt
        classData.put("student", student);
        classData.put("leader", leader);

        // return class data
        return new ResponseEntity<>(classData, HttpStatus.OK);
    }
```

## Styling updates

* New menus for creating classes

### Wireframe:

Screen Shot 2024-04-24 at 10 09 45 AM

### Integration:

If the user isn't signed in (an empty array is returned by the fetch)

Screen Shot 2024-04-24 at 10 42 45 AM

When signed in:

Screen Shot 2024-04-24 at 10 44 55 AM

### CODE REVIEW

**Comment by drewreed2005** [comment]
Quite a lovely addition to our project's functionality, Aiden. I'm quite impressed. Lovely.

File: assets/js/tablegenerator.js [hunk 1]
```diff
@@ -20,37 +26,45 @@ async function getClassList() {
             headers: {
                 "content-type": "application/json",
             },
-        });
+        })
 
+        // Error check
         if (!response.ok) {
-            throw new Error('Network response was not ok');
+            throw new Error('Network response was not ok')
```
Comment (Toby-Leeder) [inline assets/js/tablegenerator.js:33]: We should probably do this for all of our code, but whenever we throw an error you should include the status. Something more like this:

throw new Error(`HTTP error! Status: ${response.status}`);

File: assets/js/tablegenerator.js [hunk 2]
```diff
@@ -8,9 +10,13 @@ window.onload(function() {
...
 
-async function getClassList() {
+
+
+async function fetchClassList() {
+    // Create array of classes
     var classes = []
 
+    // Fetch backend
     try {
```
Comment (Toby-Leeder) [inline assets/js/tablegenerator.js:20]: Not sure if the fetch request should be in a try, catch loop unless you have a specific reason for it to be. Not 100% but it looks like it's just if you aren't signed in which is fine, but maybe there's a way to check that before the fetch request goes through so we don't miss other possible errors in the try, catch.

File: assets/js/tablegenerator.js [hunk 3]
```diff
@@ -20,37 +26,45 @@ async function getClassList() {
...
+            var studentList = []
+
+            // Push students from response into class array
+            for (let student of classData.students) {
+                studentList.push(student.name)
             }
 
-            classes.push({id: `class-${classData.id}`, class: studentList, name: classData.name});  // Ensure classData has id and name properties
+            // Add each class from response into frontend class array
+            classes.push({id: `class-${classData.id}`, class: studentList, name: classData.name})
```
Comment (Toby-Leeder) [inline assets/js/tablegenerator.js:52]: Maybe change the class element to studentList to be more clear?

File: assets/js/tablegenerator.js [hunk 4]
```diff
@@ -93,6 +107,119 @@ async function addClass() {
...
+    else {
+        classes.innerHTML = "No existing classes found, try "
+        const redirect = document.createElement("a")
+        redirect.href = `${baseurl}/sign-in`
+        redirect.innerHTML = "signing in"
+        classes.appendChild(redirect)
+        classes.innerHTML += " or creating a new class"
+    }
+
+    const createButton = document.createElement("button")
```
Comment (Toby-Leeder) [inline assets/js/tablegenerator.js:192]: Good use of document object model! No more html strings in my codebase.

File: assets/js/tablegenerator.js [hunk 5]
```diff
@@ -93,6 +107,119 @@ async function addClass() {
...
+async function addClass() {
+    const existingClasses = await fetchClassList()
+
+    console.log(existingClasses)
+
+    const main = $("#table-div")[0]
+
+    main.innerHTML = ""
+
+    const topText = document.createElement("div")
```
Comment (Toby-Leeder) [inline assets/js/tablegenerator.js:140]: topText

File: assets/js/tablegenerator.js [hunk 6]
```diff
@@ -93,6 +107,119 @@ async function addClass() {
...
+    title.appendChild(cancelButton)
+
+    const classes = document.createElement("div")
+    classes.id = "classes-div"
+
+    const textContainer = document.createElement("div")
+    textContainer.className = "text-area-container"
+    textContainer.appendChild(classes)
+
+    // i have given up on naming variables in an intuitive way, good luck soldiers!
```
Comment (Toby-Leeder) [inline assets/js/tablegenerator.js:162]: Bad

**Review by Toby-Leeder — COMMENTED** [review]
It was a little tough for me to review without knowing exactly the goal of this part of the project, but I'm not exactly sure on the frontend and backend communication flow. I see that you have the fetch class request which then pulls from the backend, but that's only called when adding a new class. I'm guessing that's how it's supposed to be but not sure. Also, I don't see when new classes generated on the frontend are added to the backend if that's a feature we want. Again, don't really know how this code works so I could just be saying nonsense just lmk.

File: assets/js/tablegenerator.js (hunk 4 above)
Comment (aidenhuynh) [reply assets/js/tablegenerator.js:192]: Thanks, Toby. It's clear you didn't look hard enough.

**Comment by aidenhuynh** [comment]
I have added an HTTP error status console logger pogger

**Pull request merged**